File: day1/part2.js
#!/usr/bin/env nodejs
fs = require('fs');

function compute(s) {
    let val = 0;
    let seen = new Set([val]);

    while (true) {
        for (const line of s.match(/[^\r\n]+/g)) {
            val += parseInt(line, 10);
            if (seen.has(val)) {
                return val;
            } else {
                seen.add(val);
            }
        }
    }
}

(function () {
    if (process.argv.length != 3) {
        console.log(`usage: ${process.argv[1]} FILENAME`);
        process.exit(1);
    }
    let contents = fs.readFileSync(process.argv[2], 'utf8');
    console.log(compute(contents));
}());

File: day1/src/bin/part2.rs
use std::env;
use std::process;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::path::Path;
use std::collections::HashSet;

fn compute(s: &str) -> i64 {
    let mut val = 0;
    let mut seen = HashSet::new();
    seen.insert(val);
    for line in s.lines().cycle() {
        val += line.parse::<i64>().unwrap();
        if seen.contains(&val) {
            return val;
        } else {
            seen.insert(val);
        }
    }
    panic!("unreachable")
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("usage {} INPUT", args[0]);
        process::exit(1);
    }

    let path = Path::new(&args[1]);
    let display = path.display();

    let mut file = match File::open(&path) {
        Err(why) => panic!("couldn't open {}: {}", display, why.description()),
        Ok(file) => file,
    };

    // Read the file contents into a string, returns `io::Result<usize>`
    let mut s = String::new();
    match file.read_to_string(&mut s) {
        Err(why) => panic!("couldn't read {}: {}", display, why.description()),
        Ok(_) => println!("{}", compute(&s)),
    }
}
